parachains/pallets/bridge-assets-transfer/src/lib.rs

```diff
@@ -0,0 +1,886 @@
+	// TODO:check-parameter - lets start with 1..1, maybe later we could extend this with BoundedVec
+	// TODO: bridged bridge-hub should have router for this
+	pub allowed_target_location: MultiLocation,
+
+	/// Fee which could be needed to pay in `bridge_location`
+	pub fee: Option<MultiAsset>,
+}
+
+impl From<BridgeConfig> for (MultiLocation, Option<MultiAsset>) {
+	fn from(bridge_config: BridgeConfig) -> (MultiLocation, Option<MultiAsset>) {
+		(bridge_config.bridge_location, bridge_config.fee)
+	}
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use crate::weights::WeightInfo;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+	use xcm::latest::Error as XcmError;
+	use xcm_builder::ExporterFor;
+	use xcm_executor::traits::TransactAsset;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// XCM sender which sends messages to the BridgeHub
+		type BridgeXcmSender: SendXcm;
+
+		/// Runtime's universal location
+		type UniversalLocation: Get<InteriorMultiLocation>;
+
+		/// Weight information for extrinsics in this pallet.
+		type WeightInfo: WeightInfo;
+
+		/// How to withdraw and deposit an asset for reserve.
+		type AssetTransactor: TransactAsset;
+
+		/// The configurable origin to allow bridges configuration management
+		type AdminOrigin: EnsureOrigin<Self::RuntimeOrigin>;
+
+		/// Required origin for asset transfer. If successful, it resolves to `MultiLocation`.
+		type TransferOrigin: EnsureOrigin<Self::RuntimeOrigin, Success = MultiLocation>;
+	}
+
+	/// Details of configured bridges which are allowed for transfer.
+	#[pallet::storage]
+	#[pallet::getter(fn bridges)]
+	pub(super) type Bridges<T: Config> = StorageMap<_, Blake2_128Concat, NetworkId, BridgeConfig>;
+
+	#[pallet::error]
+	#[cfg_attr(test, derive(PartialEq))]
+	pub enum Error<T> {
+		InvalidConfiguration,
+		InvalidAssets,
+		MaxAssetsLimitReached,
+		UnsupportedDestination,
+		BridgeCallError,
+		FailedToReserve,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub (super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Transfer was successfully entered to the system (does not mean already delivered)
+		TransferInitiated(XcmHash),
+
+		/// New bridge configuration was added
+		BridgeAdded,
+		/// Bridge configuration was removed
+		BridgeRemoved,
+		/// Bridge configuration was updated
+		BridgeUpdated,
+
+		/// Reserve asset passed
+		ReserveAssetsDeposited { from: MultiLocation, to: MultiLocation, assets: MultiAssets },
+		/// Reserve asset failed
+		FailedToReserve(XcmError),
+
+		/// Bridge transfer failed
+		BridgeCallError(SendError),
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
```

Comment:
I'm assuming that in time, we'll also add privileged dispatchables for creating assets on foreign chains. I.e Statemint sends the following message to Statemine:
```
Transact(
  Utility.batch(
    Assets::create_asset(location),
    Assets::set_metadata(name, symbols, decimals)
  )
)
```

Additionally If Snowbridge also ends up using this pallet, we'll likely want to add dispatches to do exactly the same, except the data in the Transact will be ABI-Encoded to work with the EVM. So that means Statemint will have a dependency on this crate of ours: https://github.com/Snowfork/ethabi-decode.